Duplicated aggregation produces only one result

## tl;dr

Given input `data.json`:

```
{"n": 1}
{"n": 2}
```

The single `sum` result shown below is unexpected, since in other SQL systems all three aggregation results are returned.

```
$ SUPER_VAM=1 super -c "SELECT SUM(n), SUM(n+1), SUM(n+2) FROM data.json;"
{sum:7}
```

## Details

Repro is with super commit 2231fd9. This is a simplification of the [ClickBench query 29](https://github.com/ClickHouse/ClickBench/blob/8440f516891d7b75563e895f9bd2a1b413d38970/duckdb/queries.sql#L30).

Perhaps the result is an unsurprising side effect of our not yet having support for columns with duplicate names? FWIW, if I add `AS` to each aggregation to assign unique column names, it works as expected.

```
$ super -version
Version: v1.18.0-445-g2231fd961

$ SUPER_VAM=1 super -c "SELECT SUM(n) AS n_0, SUM(n+1) AS n_1, SUM(n+2) AS n_2 FROM data.json;"
{n_0:3,n_1:5,n_2:7}
```

That said, other SQL systems just create unique column names in this situation. Examples:

```
$ duckdb -c "SELECT SUM(n), SUM(n+1), SUM(n+2) FROM data.json;"
┌────────┬──────────────┬──────────────┐
│ sum(n) │ sum((n + 1)) │ sum((n + 2)) │
│ int128 │    int128    │    int128    │
├────────┼──────────────┼──────────────┤
│   3    │      5       │      7       │
└────────┴──────────────┴──────────────┘

$ clickhouse --output-format TabSeparatedWithNames --query "SELECT SUM(n), SUM(n+1), SUM(n+2) FROM 'data.json';"
SUM(n)	SUM(plus(n, 1))	SUM(plus(n, 2))
3	5	7
```

Whereas in SuperDB"s vector runtime as shown above it looks like only the "rightmost" aggregation result is preserved.

FWIW, in sequential runtime, it's a panic instead.

```
$ super -c "SELECT SUM(n), SUM(n+1), SUM(n+2) FROM data.json;"
panic: duplicate field: "sum"
goroutine 1 [running]:
runtime/debug.Stack()
	/usr/local/opt/go/libexec/src/runtime/debug/stack.go:26 +0x5e
github.com/brimdata/super/runtime/sam/op.(*Catcher).Pull.func1()
	/Users/phil/work/super/runtime/sam/op/catcher.go:25 +0x3d
panic({0x97bc860?, 0xc0004a4320?})
	/usr/local/opt/go/libexec/src/runtime/panic.go:792 +0x132
github.com/brimdata/super.(*Context).MustLookupTypeRecord(...)
	/Users/phil/work/super/context.go:179
github.com/brimdata/super/runtime/sam/expr.(*recordExpr).Eval(0xc0006157c0, {0x51bf7360, 0xc0000a6d50}, {{0x9a8ca60?, 0xc0000a6cc0?}, 0xc0004ae918?, 0x20?})
	/Users/phil/work/super/runtime/sam/expr/values.go:60 +0x3b3
github.com/brimdata/super/runtime/sam/op/yield.(*Op).Pull(0xc00047e840, 0x0)
	/Users/phil/work/super/runtime/sam/op/yield/yield.go:34 +0x42b
github.com/brimdata/super/runtime/sam/op.(*Single).Pull(0xc0000a6ae0, 0x5?)
	/Users/phil/work/super/runtime/sam/op/mux.go:120 +0x33
github.com/brimdata/super/runtime/sam/op.(*Catcher).Pull(0x4e62d91f42a2d62a?, 0x39?)
	/Users/phil/work/super/runtime/sam/op/catcher.go:28 +0x5c
github.com/brimdata/super/runtime/exec.(*Query).Pull(0x7adf0df?, 0x60?)
	/Users/phil/work/super/runtime/exec/query.go:49 +0x3c
github.com/brimdata/super/zbuf.CopyMux(0xc00059fca8, {0x9a83a00, 0xc0000a6b10})
	/Users/phil/work/super/zbuf/mux.go:40 +0x38
github.com/brimdata/super/cmd/super/root.(*Command).Run(0xc00042e600, {0xc00003c710, 0x0, 0x0})
	/Users/phil/work/super/cmd/super/root/command.go:167 +0x925
github.com/brimdata/super/pkg/charm.path.run({0xc000080c08, 0x1, 0x1}, {0xc00003c710, 0x0, 0x0?})
	/Users/phil/work/super/pkg/charm/path.go:11 +0x7b
github.com/brimdata/super/pkg/charm.(*Spec).Exec(0xabd2f20, {0xc00003c700, 0x2, 0x2})
	/Users/phil/work/super/pkg/charm/charm.go:71 +0x105
main.main()
	/Users/phil/work/super/cmd/super/main.go:40 +0x5b
```

And in vector runtime, this changed as of the merge of #5669. Before that, there also was an error message in vector runtime. As of the commit right before then:

```
$ super -version
Version: v1.18.0-293-g0f8e7dbef

$ SUPER_VAM=1 super -c "SELECT SUM(n), SUM(n+1), SUM(n+2) FROM data.json;"
duplicate field: "sum"
```

--- a/compiler/semantic/sql.go
+++ b/compiler/semantic/sql.go
@@ -533,7 +533,7 @@ func (a *analyzer) semGroupBy(sch *selectSchema, in []ast.Expr) []exprloc {
 func (a *analyzer) semProjection(sch *selectSchema, args []ast.AsExpr, funcs *aggfuncs) projection {
 	out := &anonSchema{}
 	sch.out = out
-	conflict := make(map[string]struct{})
+	aliases := make(map[string]struct{})
 	var proj projection
 	for _, as := range args {
 		if isStar(as) {
@@ -541,19 +541,42 @@ func (a *analyzer) semProjection(sch *selectSchema, args []ast.AsExpr, funcs *ag
 			continue
 		}
 		col := a.semAs(sch, as, funcs)
-		//XXX check for conflict for now, but we're getting rid of this soon
-		if _, ok := conflict[col.name]; ok {
-			n := ast.Node(as.Expr)
-			if as.ID != nil {
-				n = as.ID
+		if as.ID != nil {
+			if _, ok := aliases[col.name]; ok {
+				a.error(as.ID, fmt.Errorf("duplicate column label %q", col.name))
+				continue
 			}
-			a.error(n, fmt.Errorf("%q: conflicting name in projection; try an AS clause", col.name))
+			aliases[col.name] = struct{}{}
 		}
-		conflict[col.name] = struct{}{}
 		proj = append(proj, *col)
+	}
+	// Do we have duplicates for inferred columns? If an alias and an
+	// inferred column collide the inferred name should be changed.
+	for i := range proj {
+		col := &proj[i]
+		if col.isStar() {
+			continue
+		}
+		if args[i].ID == nil {
+			col.name = dedupeColname(aliases, col.name)
+		}
+		aliases[col.name] = struct{}{}
 		out.columns = append(out.columns, col.name)
 	}
+
 	return proj
+}
+
+func dedupeColname(m map[string]struct{}, name string) string {
+	for i := 0; ; i++ {
+		s := name
+		if i > 0 {
+			s += fmt.Sprintf("_%d", i)
+		}
+		if _, ok := m[s]; !ok {
+			return s
+		}
+	}
 }
 
 func isStar(a ast.AsExpr) bool {
@@ -591,11 +614,14 @@ func (a *analyzer) semExprSchema(s schema, e ast.Expr) dag.Expr {
 // that's what SQL does!  And it seems different implementations format
 // expressions differently.  XXX we need to check ANSI SQL spec here
 func inferColumnName(e dag.Expr, ae ast.Expr) string {
-	path, err := deriveLHSPath(e)
-	if err != nil {
+	switch e := e.(type) {
+	case *dag.This:
+		return field.Path(e.Path).Leaf()
+	case *dag.Var:
+		return e.Name
+	default:
 		return zfmt.ASTExpr(ae)
 	}
-	return field.Path(path).Leaf()
 }
 
 func yieldExpr(e dag.Expr, seq dag.Seq) dag.Seq {

--- a/compiler/ztests/sql/agg-distinct.yaml
+++ b/compiler/ztests/sql/agg-distinct.yaml
@@ -26,5 +26,5 @@ input: |
   {key:"a",n:1,b:false}
 
 output: |
-  {key:"a",and:false,any:1,avg:1.5,collect:["a"],count:2(uint64),dcount:2(uint64),max:2,min:1,or:true,sum:3,union:|[1,2]|}
-  {key:"b",and:true,any:1,avg:0.5,collect:["b"],count:2(uint64),dcount:2(uint64),max:1,min:0,or:true,sum:1,union:|[0,1]|}
+  {key:"a","and(distinct b)":false,"any(distinct n)":1,"avg(distinct n)":1.5,"collect(distinct key)":["a"],"count(distinct n)":2(uint64),"dcount(distinct n)":2(uint64),"max(distinct n)":2,"min(distinct n)":1,"or(distinct b)":true,"sum(distinct n)":3,"union(distinct n)":|[1,2]|}
+  {key:"b","and(distinct b)":true,"any(distinct n)":1,"avg(distinct n)":0.5,"collect(distinct key)":["b"],"count(distinct n)":2(uint64),"dcount(distinct n)":2(uint64),"max(distinct n)":1,"min(distinct n)":0,"or(distinct b)":true,"sum(distinct n)":1,"union(distinct n)":|[0,1]|}

--- a/compiler/ztests/sql/anycase-agg-func.yaml
+++ b/compiler/ztests/sql/anycase-agg-func.yaml
@@ -12,4 +12,4 @@ inputs:
 outputs:
   - name: stdout
     data: |
-      {count:4(uint64),sum:10}
+      {"COUNT()":4(uint64),"Sum(c)":10}

--- a/compiler/ztests/sql/duplicate-as.yaml
+++ b/compiler/ztests/sql/duplicate-as.yaml
@@ -4,6 +4,6 @@ script: |
 outputs:
   - name: stderr
     data: |
-      "foo": conflicting name in projection; try an AS clause at line 1, column 23:
+      duplicate column label "foo" at line 1, column 23:
       SELECT 1 as foo, 2 as foo
                             ~~~

--- a/compiler/ztests/sql/group-by-expr.yaml
+++ b/compiler/ztests/sql/group-by-expr.yaml
@@ -17,17 +17,17 @@ inputs:
 outputs:
   - name: stdout
     data: |
-      {"c+a":12,collect:[1]}
-      {"c+a":14,collect:[2,1]}
-      {"c+a":16,collect:[2]}
+      {"c+a":12,"collect(c)":[1]}
+      {"c+a":14,"collect(c)":[2,1]}
+      {"c+a":16,"collect(c)":[2]}
       ===
-      {d:12,collect:[1]}
-      {d:14,collect:[2,1]}
-      {d:16,collect:[2]}
+      {d:12,"collect(c)":[1]}
+      {d:14,"collect(c)":[2,1]}
+      {d:16,"collect(c)":[2]}
       ===
-      {d:12,collect:[1]}
-      {d:14,collect:[2,1]}
-      {d:16,collect:[2]}
+      {d:12,"collect(c)":[1]}
+      {d:14,"collect(c)":[2,1]}
+      {d:16,"collect(c)":[2]}
   - name: err1
     data: |
       no corresponding grouping element for non-aggregate "c+a" at line 1, column 8:

--- a/compiler/ztests/sql/groupby.yaml
+++ b/compiler/ztests/sql/groupby.yaml
@@ -14,10 +14,10 @@ inputs:
 outputs:
   - name: stdout
     data: |
-      {radius:1.5,count:1(uint64)}
-      {radius:0.7,count:1(uint64)}
-      {radius:4.2,count:1(uint64)}
-      {radius:error("missing"),count:1(uint64)}
+      {radius:1.5,"count()":1(uint64)}
+      {radius:0.7,"count()":1(uint64)}
+      {radius:4.2,"count()":1(uint64)}
+      {radius:error("missing"),"count()":1(uint64)}
       ===
-      {type:"circle",sum:6.4}
-      {type:"rectangle",sum:null}
+      {type:"circle","sum(val.radius)":6.4}
+      {type:"rectangle","sum(val.radius)":null}

--- a/compiler/ztests/sql/having-no-group.yaml
+++ b/compiler/ztests/sql/having-no-group.yaml
@@ -12,4 +12,4 @@ inputs:
 outputs:
   - name: stdout
     data: |
-      {sum:6}
+      {"sum(b)":6}

--- a/compiler/ztests/sql/implied-as.yaml
+++ b/compiler/ztests/sql/implied-as.yaml
@@ -2,6 +2,8 @@ script: |
   super -z -c "select a x from t.json"
   echo ===
   super -z -c "select a x, a as y, b z from t.json"
+  echo ===
+  super compile -C -O 'select n+1, n+2, n+1, n+3, n+1'
 
 inputs:
   - name: t.json
@@ -23,3 +25,7 @@ outputs:
       {x:2,y:2,z:1}
       {x:3,y:3,z:2}
       {x:4,y:4,z:2}
+      ===
+      null
+      | yield {"n+1":n+1,"n+2":n+2,"n+1_1":n+1,"n+3":n+3,"n+1_2":n+1}
+      | output main

--- a/compiler/ztests/sql/semicolon.yaml
+++ b/compiler/ztests/sql/semicolon.yaml
@@ -11,4 +11,4 @@ inputs:
 outputs:
   - name: stdout
     data: |
-      {count:3(uint64),sum:6}
+      {"count(a)":3(uint64),"sum(a)":6}

--- a/runtime/ztests/expr/function/length.yaml
+++ b/runtime/ztests/expr/function/length.yaml
@@ -6,4 +6,4 @@ input: |
   null
 
 output: |
-  {length:13}
+  {"LENGTH(\"W3Schools.com\")":13}

